pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java

```diff
@@ -2477,6 +2478,7 @@ public void startSendOperation(Producer producer, int msgSize, int numMessages)
                     producer.getTopic().isTopicPublishRateExceeded(numMessages, msgSize);
             if (isPreciseTopicPublishRateExceeded) {
                 producer.getTopic().disableCnxAutoRead();
+                recordRateLimitMetrics(producers.values());
```

Comment:
> producer.getTopic().disableCnxAutoRead();

This is going to throttle all the connections with producers of this `producer.getTopic()`. 
We need to `recordRateLimitMetrics` on the topics in these connections too.

For example.
Connection (1) have Producer of topic A and B.
Connection (2) have Producer of topic A and C.
One send request in Connection (1) is causing topic A under throttling. 
And Topic B and C is also under throttling because they share the same connection.
And Topic C won't be called `increasePublishLimitedTimes` with current code.

Correct me if I miss something.

Reply:
Each `ServerCnx` contains multiple Producers (all Producers that share this channel). I have traversed all the Producers here, and each Producer has its corresponding Topic, so all Topics that share this connection will be  +1.
In the scenario you describe, the current limit times of Topic C will also be +1

Reply:
> Each `ServerCnx` contains multiple Producers (all Producers that share this channel). I have traversed all the Producers here, and each Producer has its corresponding Topic, so all Topics that share this connection will be +1.

Yes, that's why topic B will +1. 

My scenario is that `startSendOperation` is only called in `ServerCnx1`. 
`producers.values()` in Line 2481 only contains producers of Topic A and B (They shares the same connection `ServerCnx1` ).

But `startSendOperation` is not called on  `ServerCnx2`, so the rate limit times of Topic C won't be +1, right?

And in fact `ServerCnx2` is still rate limitted due to  this code path :
`producer.getTopic().disableCnxAutoRead();` -->  `producers.values().forEach(producer -> producer.getCnx().disableCnxAutoRead());` (Topic A  has producer in ServerCnx2)

Reply:
`ServerCnx2` is another channel, connections bound to `ServerCnx1` will not effect connections bound to `ServerCnx2`. `producers.values()` only get the producers belong to `ServerCnx1`.

So the code path you show will only disable auto read for connections bound to `ServerCnx1`.  Connections bound to `ServerCnx2` can go on producing.

```
Connection (1) have Producer of topic A and B.
Connection (2) have Producer of topic A and C.
```
If Topic A is a multi-partition topic, only some partitions will be throttled.

Reply:
> `producers.values().forEach(producer -> producer.getCnx().disableCnxAutoRead());`

This line is in `org.apache.pulsar.broker.service.AbstractTopic#disableCnxAutoRead`,  the `producers` is  `AbstractTopic#producers` not `ServerCnx#producers`.
So it contains producers in both `ServerCnx1` and `ServerCnx2`

Reply:
`protected final ConcurrentHashMap<String, Producer> producers;`
producers in AbstractTopic is not static, it only contains the producers belong to this  topic/partition
Different topics will have different AbstractTopic objects